### PR TITLE
[F# 4.2] Report better error for invalid = in for loop

### LINKED ISSUE #1711: Improve error reporting: suggest using `in` instead of mistaken `=`
# what

when writing for loop

```fsharp
for i = 0 .. 100 do
    ()
```

FSI yields

> Unexpected symbol '..' in expression

# why

I made a mistake using `=` instead of `in`

# how

I'd like the compiler to show me how to fix:

> Unexpected symbol '..' in expression, were you intending to use for x in y .. z do?

Related #1430

### PR DESCRIPTION
**Pull request by forki**
fixes #1711

### CODE REVIEW

**Comment by forki** [comment]
ready for review

**Comment by forki** [comment]
is this error a false positive?

**Comment by dsyme** [comment]
@forki That CI error looks unrelated, and inconsistent.  I've not seen that CoreCLR test fail before

File: src/fsharp/FSComp.txt [hunk 1]
```diff
@@ -1328,3 +1328,4 @@ tcTupleStructMismatch,"One tuple type is a struct tuple, the other is a referenc
 3210,tcNamedActivePattern,"%s is an active pattern and cannot be treated as a discriminated union case with named fields."
 3211,DefaultParameterValueNotAppropriateForArgument,"The default value does not have the same type as the argument. The DefaultParameterValue attribute and any Optional attribute will be ignored. Note: 'null' needs to be annotated with the correct type, e.g. 'DefaultParameterValue(null:obj)'."
 tcGlobalsSystemTypeNotFound,"The system type '%s' was required but no referenced system DLL contained this type"
+3213,parsUnexpectedSymbolEqualsInsteadOfIn,"Unexpected symbol '=' in expression, were you intending to use 'for x in y .. z do' instead?"
```
Comment (dsyme) [inline src/fsharp/FSComp.txt]: Please use two sentences, along the lines "This is the error.  This is the analysis. This is the advice." thanks

**Comment by dsyme** [comment]
@forki Because of the change to the parser I will mark this as F# 4.2

**Comment by KevinRansom** [comment]
I like the error message improvement.
@dsyme this isn't a language change Don, and the change to the parser is very localized. I think we can take this now rather than waiting.

**Comment by forki** [comment]
Sorry. Will change that tomorrow.

Am 05.12.2016 11:57 nachm. schrieb "Jared Hester" <notifications@github.com

File: src/fsharp/FSComp.txt [hunk 2]
```diff
@@ -1328,3 +1328,4 @@ tcTupleStructMismatch,"One tuple type is a struct tuple, the other is a referenc
 3210,tcNamedActivePattern,"%s is an active pattern and cannot be treated as a discriminated union case with named fields."
 3211,DefaultParameterValueNotAppropriateForArgument,"The default value does not have the same type as the argument. The DefaultParameterValue attribute and any Optional attribute will be ignored. Note: 'null' needs to be annotated with the correct type, e.g. 'DefaultParameterValue(null:obj)'."
 tcGlobalsSystemTypeNotFound,"The system type '%s' was required but no referenced system DLL contained this type"
+3213,parsUnexpectedSymbolEqualsInsteadOfIn,"Unexpected symbol '=' in expression. Were you intending to use 'for x in y .. z do' instead?"
```
Comment (KevinRansom) [inline src/fsharp/FSComp.txt]: It is "past perfect progressive."

Might I suggest:
````
Consider: 'for x in y .. z do' instead?
````

**Comment by forki** [comment]
fixed

**Comment by dsyme** [comment]
@KevinRansom Parser changes always have risks, because the introduction of new conflicts can potentially change how other good-but-obscure programs get parsed.  We have to tread carefully with those.  Since there's a risk, and having a parser bug in RTM would be really bad, I think we should hold off.

**Comment by forki** [comment]
I think 4.2 is absolutely fine.

**Comment by KevinRansom** [comment]
@dsyme

We are heading for a faster release model I want to pull this into master to get as much coverage as we can before the dotnet framework 2.0 release.

**Comment by KevinRansom** [comment]
@forki, 1 test failed, it seems relevant:

````
+++ Warnings (EqualsInsteadOfInInForLoop.fs) +++
Microsoft (R) F# Compiler version 4.1
Copyright (c) Microsoft Corporation. All Rights Reserved.
(4,7): error FS3215: Unexpected symbol '=' in expression. Did you intend to use 'for x in y .. z do' instead?

*** The following necessary lines were never matched:
***	error FS3213:.+Unexpected symbol '=' in expression. Did you intend to use 'for x in y .. z do' instead?


*** The following necessary lines were incorrectly matched:

Unexpected Compiler Output 
FAIL
````

**Comment by KevinRansom** [comment]
@forki 

Thanks for taking care of this

Kevin

**Pull request merged**